Sources/FoundationInternationalization/Formatting/Date/Date+RelativeFormatStyle.swift

```diff
@@ -151,69 +212,108 @@ extension Date {
             let refDateEnd = refDateStart.addingTimeInterval(interval - 1)
             let dateComponents: DateComponents
             if refDate < destDate {
-                dateComponents = calendar.dateComponents(Set(ICURelativeDateFormatter.sortedAllowedComponents), from: refDateStart, to: destDate)
+                dateComponents = calendar.dateComponents(allowedComponents, from: refDateStart, to: destDate)
             } else {
-                dateComponents = calendar.dateComponents(Set(ICURelativeDateFormatter.sortedAllowedComponents), from: refDateEnd, to: destDate)
+                dateComponents = calendar.dateComponents(allowedComponents, from: refDateEnd, to: destDate)
             }
 
             return dateComponents.nonZeroComponentsAndValue.first
         }
 
-        private static func _roundedLargestComponentValue(components: DateComponents, for destDate: Date, calendar: Calendar) -> CalendarComponentAndValue? {
+        private static func _roundedLargestComponentValue(refDate: Date, for destDate: Date, calendar: Calendar, allowedComponents: Set<Calendar.Component>, largestAllowedComponent: Calendar.Component) -> CalendarComponentAndValue? {
             // "day: 1, hour: 11" -> "day: 1"
             // "day: 1, hour: 15" -> "day: 2"
             // "hour: 23, minute: 50" -> "hour: 24" // only carry to the immediate previous value
             // "hour: -23, minute: -30" -> "hour: -24"
 
-            let compsAndValues = components.nonZeroComponentsAndValue
+            let components = calendar.dateComponents(Set(ICURelativeDateFormatter.sortedAllowedComponents.filter({
+                Date.ComponentsFormatStyle.Field.Option(component: $0)! <= Date.ComponentsFormatStyle.Field.Option(component: largestAllowedComponent)!
+            })).union([.nanosecond]), from: refDate, to: destDate)
+
+            let largestNonZeroComponent = components.nonZeroComponentsAndValue.first ?? (.nanosecond, components.nanosecond ?? 0)
```

Comment:
This variable name is misleading -- it's still possible that this returns some date components with nanoseconds == 0

Reply:
Fixed it so it's nil if it would have been nanosecond before. Actually made that part a lot cleaner, thanks!